// @microsoft/sp-http/lib/oauthTokenProvider/OAuth2TokenProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var sp_telemetry_1 = require("@ms/sp-telemetry");
var decorators_1 = require("@microsoft/decorators");
var sp_core_library_1 = require("@microsoft/sp-core-library");
var OAuth2Constants_1 = require("./OAuth2Constants");
var OAuth2TokenProvider = (function () {
    function OAuth2TokenProvider(serviceScope) {
        this._displayCallHandler = undefined;
        this._loadAndConfigureAdalJsModulePromise = undefined;
        this._loginUserPromise = undefined;
    }
    OAuth2TokenProvider_1 = OAuth2TokenProvider;
    OAuth2TokenProvider.prototype._initialize = function (azureInstanceUrl, azureTenantId) {
        sp_core_library_1.Validate.isNonemptyString(azureInstanceUrl, 'azureInstanceUrl');
        sp_core_library_1.Validate.isNonemptyString(azureTenantId, 'azureTenantId');
        this._azureInstanceUrl = azureInstanceUrl;
        this._azureTenantId = azureTenantId;
        this._isInitialized = true;
    };
    OAuth2TokenProvider.prototype.getOAuthToken = function (resourceEndpoint) {
        var _this = this;
        if (!this._isInitialized) {
            return Promise.reject(new Error('This operation cannot be performed until the OAuth2TokenProvider is initialized.'));
        }
        if (!sp_core_library_1._SPFlight.isEnabled(375 )) {
            return Promise.reject(new Error('The requested operation is part of an experimental feature ' +
                'that is not supported in the current environment.'));
        }
        return this._loginUser().then(function () {
            return _this._acquireTokenPromise(resourceEndpoint).then(function (token) {
                return token.token;
            });
        });
    };
    OAuth2TokenProvider.prototype._acquireTokenPromise = function (resourceEndpoint) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var acquireAccessTokenQosMonitor = new sp_telemetry_1._QosMonitor('OAuth2TokenProvider.AcquireAccessToken');
            OAuth2TokenProvider_1._authContext.acquireToken(resourceEndpoint, function (message, token) {
                if (!token) {
                    _this._acquireTokenPopupPromise(resourceEndpoint)
                        .then(function (popUpToken) {
                        if (popUpToken.message) {
                            acquireAccessTokenQosMonitor.writeSuccess({ 'Message': popUpToken.message });
                        }
                        else {
                            acquireAccessTokenQosMonitor.writeSuccess();
                        }
                        resolve(popUpToken);
                    })
                        .catch(function (e) {
                        acquireAccessTokenQosMonitor.writeUnexpectedFailure('AcquireAccessToken' , e);
                        reject(e);
                    });
                }
                else {
                    if (message) {
                        acquireAccessTokenQosMonitor.writeSuccess({ 'Message': message });
                    }
                    else {
                        acquireAccessTokenQosMonitor.writeSuccess();
                    }
                    resolve({
                        message: message,
                        token: token
                    });
                }
            });
        });
    };
    OAuth2TokenProvider.prototype._acquireTokenPopupPromise = function (resourceEndpoint) {
        return new Promise(function (resolve, reject) {
            OAuth2TokenProvider_1._authContext.acquireTokenPopup(resourceEndpoint, undefined , undefined , function (message, token) {
                if (!token) {
                    reject(new Error(message));
                }
                resolve({
                    message: message,
                    token: token
                });
            });
        });
    };
    OAuth2TokenProvider.prototype._loadAndConfigureAdalJsModule = function () {
        var _this = this;
        if (!this._loadAndConfigureAdalJsModulePromise) {
            this._loadAndConfigureAdalJsModulePromise = new Promise(function (resolve, reject) {
                try {
                    require.ensure(['adal-angular'], function (require) {
                        resolve(require('adal-angular'));
                    }, 'sp-http-adal');
                }
                catch (e) {
                    reject(e);
                }
            }).then(function (adalJsModule) {
                _this._configureAdalJs(adalJsModule);
                return adalJsModule;
            });
        }
        return this._loadAndConfigureAdalJsModulePromise;
    };
    OAuth2TokenProvider.prototype._configureAdalJs = function (adalJsModule) {
        var _this = this;
        var redirectUri = window.location.origin + '/_forms/singlesignon.aspx';
        OAuth2TokenProvider_1._authContext = adalJsModule.inject({
            clientId: OAuth2Constants_1.default.SPO_WEB_APP_PRINCIPAL_ID,
            redirectUri: redirectUri,
            instance: this._azureInstanceUrl + '/',
            tenant: this._azureTenantId,
            popUp: true,
            displayCall: function (urlNavigate) {
                if (_this._displayCallHandler) {
                    _this._displayCallHandler({
                        urlNavigate: urlNavigate,
                        redirectUri: redirectUri
                    });
                }
            }
        });
    };
    OAuth2TokenProvider.prototype._startLogin = function (showWindowCallback) {
        if (this._displayCallHandler) {
            return; 
        }
        this._displayCallHandler = function (args) {
            showWindowCallback(args);
        };
        OAuth2TokenProvider_1._authContext._loginInProgress = false;
        OAuth2TokenProvider_1._authContext.login();
        this._displayCallHandler = undefined;
        return;
    };
    OAuth2TokenProvider.prototype._loginUser = function () {
        var _this = this;
        if (this._loginUserPromise) {
            return this._loginUserPromise;
        }
        return this._loadAndConfigureAdalJsModule()
            .then(function (adalJsModule) {
            sp_core_library_1.Validate.isNotNullOrUndefined(OAuth2TokenProvider_1._authContext, 'AuthenticationContext');
            if (OAuth2TokenProvider_1._authContext.getCachedToken(OAuth2Constants_1.default.SPO_WEB_APP_PRINCIPAL_ID)) {
                OAuth2TokenProvider_1._authContext.getCachedUser();
                return; 
            }
            else {
                var acquireIdTokenMonitor_1 = new sp_telemetry_1._QosMonitor('OAuth2TokenProvider.AcquireIdToken');
                var acquireIdTokenTagNameSuffix_1 = 'AcquireIdToken';
                _this._startLogin(function (args) {
                    var popupWindow = window.open(args.urlNavigate, 'login', 'width=483, height=600');
                    if (!popupWindow) {
                        var popUpDidNotOpenError = new Error('Failed to open pop-up window');
                        acquireIdTokenMonitor_1.writeExpectedFailure(acquireIdTokenTagNameSuffix_1, popUpDidNotOpenError);
                        throw popUpDidNotOpenError;
                    }
                    if (popupWindow.closed === undefined) {
                        var popUpClosedPropertyMissingError = new Error('This operation requires the popupWindow.closed property, which is not supported in this browser');
                        acquireIdTokenMonitor_1.writeExpectedFailure(acquireIdTokenTagNameSuffix_1, popUpClosedPropertyMissingError);
                        throw popUpClosedPropertyMissingError;
                    }
                    if (popupWindow.focus) {
                        popupWindow.focus();
                    }
                    _this._loginUserPromise = _this._waitForWindowToClose(popupWindow, args)
                        .then(function () {
                        acquireIdTokenMonitor_1.writeSuccess();
                        _this._loginUserPromise = undefined;
                    }).catch(function (e) {
                        acquireIdTokenMonitor_1.writeExpectedFailure(acquireIdTokenTagNameSuffix_1, e);
                        _this._loginUserPromise = undefined;
                        throw e;
                    });
                });
                if (!_this._loginUserPromise) {
                    var unreachableError = new Error('Should be impossible.');
                    acquireIdTokenMonitor_1.writeUnexpectedFailure(acquireIdTokenTagNameSuffix_1, unreachableError);
                    throw unreachableError;
                }
                return _this._loginUserPromise;
            }
        });
    };
    OAuth2TokenProvider.prototype._waitForWindowToClose = function (popupWindow, args) {
        return new Promise(function (resolve, reject) {
            var WINDOW_POLL_MS = 100;
            var ATTEMPTS_TO_POLL_WINDOW = !DEPRECATED_UNIT_TEST ? 50 : 1;
            try {
                var currentNumberOfPollAttempts_1 = 0;
                var pollTimer_1 = window.setInterval(function () {
                    if (currentNumberOfPollAttempts_1 === ATTEMPTS_TO_POLL_WINDOW) {
                        window.clearInterval(pollTimer_1);
                        popupWindow.close();
                        reject(new Error('Login popup failed to redirect to the designated redirect uri.'));
                        return;
                    }
                    if (popupWindow.closed) {
                        window.clearInterval(pollTimer_1);
                        reject(new Error('Window closed without authenticating'));
                        return;
                    }
                    if (popupWindow.document.URL.indexOf(args.redirectUri) !== -1) {
                        window.clearInterval(pollTimer_1);
                        if (!popupWindow.location.hash) {
                            popupWindow.close();
                            reject(new Error('Login popup redirected to the redirect uri but no id token was found.'));
                            return;
                        }
                        OAuth2TokenProvider_1._authContext.handleWindowCallback(popupWindow.location.hash);
                        popupWindow.close();
                        resolve();
                        return;
                    }
                    currentNumberOfPollAttempts_1++;
                }, WINDOW_POLL_MS);
            }
            catch (e) {
                reject(e);
            }
        });
    };
    OAuth2TokenProvider.serviceKey = sp_core_library_1.ServiceKey.create('sp-http:OAuth2TokenProvider', OAuth2TokenProvider_1);
    OAuth2TokenProvider = OAuth2TokenProvider_1 = tslib_1.__decorate([
        decorators_1.sealed
    ], OAuth2TokenProvider);
    return OAuth2TokenProvider;
    var OAuth2TokenProvider_1;
}());
exports.default = OAuth2TokenProvider;
